drivers/crypto/crypto_stm32.c

```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright (c) 2020 Markus Fuchs <markus.fuchs@de.sauter-bc.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <kernel.h>
+#include <device.h>
+#include <assert.h>
+#include <crypto/cipher.h>
+#include <clock_control/stm32_clock_control.h>
+#include <drivers/clock_control.h>
+#include <sys/byteorder.h>
+
+#include "crypto_stm32_priv.h"
+
+#define LOG_LEVEL CONFIG_CRYPTO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(crypto_stm32);
+
+#define CRYP_SUPPORT (CAP_RAW_KEY | CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS)
+#define BLOCK_LEN_BYTES 16
+#define BLOCK_LEN_WORDS (BLOCK_LEN_BYTES / sizeof(u32_t))
+#define CRYPTO_MAX_SESSION CONFIG_CRYPTO_STM32_MAX_SESSION
+
+struct crypto_stm32_session crypto_stm32_sessions[CRYPTO_MAX_SESSION];
+
+static void copy_reverse_words(u8_t *dst_buf, int dst_len,
+			       u8_t *src_buf, int src_len)
+{
+	int i;
+
+	__ASSERT_NO_MSG(dst_len >= src_len);
+	__ASSERT_NO_MSG((dst_len % 4) == 0);
+
+	memcpy(dst_buf, src_buf, src_len);
+	for (i = 0; i < dst_len; i += sizeof(u32_t)) {
+		sys_mem_swap(&dst_buf[i], sizeof(u32_t));
+	}
+}
+
+static int do_encrypt(struct cipher_ctx *ctx, u8_t *in_buf, int in_len,
+		      u8_t *out_buf)
+{
+	HAL_StatusTypeDef status;
+
+	struct crypto_stm32_data *data = CRYPTO_STM32_DATA(ctx->device);
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	status = HAL_CRYP_SetConfig(&data->hcryp, &session->config);
```

Comment:
I would have expected HAL_CRYP_SetConfig() outside do_encrypt() because of setting once should be enough.

Reply:
I wanted to avoid code duplication, so I moved `HAL_CRYP_SetConfig` from the `crypto_stm32_XXX_XXcrypt` functions to `do_XXcrypt`. Moving it outside the crypt functions wouldn't work well in a multi-session scenario, I guess.

Reply:
Ok.